src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/SignatureHelper.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+
+namespace System.Reflection.Emit.Experiment
+{
+    // This is prototype code, to generate simple signatures.
+    // For more complex signatures, will port System.Reflection.Emit's SignatureHelper.
+    internal static class MetadataSignatureHelper
+    {
+        internal static BlobBuilder FieldSignatureEncoder(Type fieldType)
+        {
+            var fieldSignature = new BlobBuilder();
+
+            MapReflectionTypeToSignatureType(new BlobEncoder(fieldSignature).FieldSignature(), fieldType);
+
+            return fieldSignature;
+        }
+
+        internal static BlobBuilder MethodSignatureEncoder(ParameterInfo[]? parameters, ParameterInfo? returnType, bool isInstance)
+        {
+            Type[]? typeParameters = null;
+            Type? typeReturn = null;
+
+            if (parameters != null)
+            {
+                typeParameters = Array.ConvertAll(parameters, parameter => parameter.ParameterType);
+            }
+
+            if (returnType != null)
+            {
+                typeReturn = returnType.ParameterType;
+            }
+
+            return MethodSignatureEncoder(typeParameters, typeReturn, isInstance);
+        }
+        internal static BlobBuilder MethodSignatureEncoder(Type[]? parameters, Type? returnType, bool isInstance)
+        {
+            // Encoding return type and parameters.
+            var methodSignature = new BlobBuilder();
+
+            ParametersEncoder _parEncoder;
+            ReturnTypeEncoder _retEncoder;
+
+            new BlobEncoder(methodSignature).
+                MethodSignature(isInstanceMethod: isInstance).
+                Parameters((parameters == null) ? 0 : parameters.Length, out _retEncoder, out _parEncoder);
+
+            if (returnType != null && returnType != typeof(void))
+            {
+                MapReflectionTypeToSignatureType(_retEncoder.Type(), returnType);
+            }
+            else // If null mark ReturnTypeEncoder as void
+            {
+                _retEncoder.Void();
+            }
+
+            if (parameters != null) // If parameters null, just keep empty ParametersEncoder empty
+            {
+                foreach (var parameter in parameters)
+                {
+                    MapReflectionTypeToSignatureType(_parEncoder.AddParameter().Type(), parameter);
+                }
+            }
+            return methodSignature;
+        }
+
+        private static void MapReflectionTypeToSignatureType(SignatureTypeEncoder signature, Type type)
```

Comment:
Nit: I would expect method that is called "Map..." to return the mapped value. Either change the name of the method or what it does.